rfcs/45-stageless.md

```diff
@@ -0,0 +1,977 @@
+    systems: Vec<RefCell<BoxedSystem>>,
+    system_conditions: Vec<RefCell<Vec<BoxedRunCondition>>>,
+    set_conditions: Vec<RefCell<Vec<BoxedRunCondition>>>,
+    // for returning things back to `Systems`
+    system_ids: Vec<NodeId>,
+    set_ids: Vec<NodeId>,
+    // for obeying dependencies
+    system_deps: Vec<(usize, Vec<usize>)>,
+    // for evaluating set conditions and skipping systems
+    sets_of_systems: Vec<FixedBitSet>,
+    sets_of_sets: Vec<FixedBitSet>,
+    systems_of_sets: Vec<FixedBitSet>,
+}
+```
+
+### Using descriptor API for both systems and system sets
+
+It's fairly straightforward to register pure labels as system sets using the same API.
+
+As a bonus, schedule construction will become completely order-independent as everything will be deferred until the end.
+
+```rust
+pub trait IntoConfiguredSystem<Params> {
+    fn configure(self) -> ConfiguredSystem;
+    fn configure_with(self, config: Config) -> ConfiguredSystem;
+    fn before<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSystem;
+    fn after<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSystem;
+    fn in_set(self, set: impl SystemLabel) -> ConfiguredSystem;
+    fn run_if<P>(self, condition: impl IntoRunCondition<P>) -> ConfiguredSystem;
+    fn named(self, name: impl SystemLabel) -> ConfiguredSystem;
+}
+
+pub trait IntoConfiguredSystemSet {
+    fn configure(self) -> ConfiguredSystemSet;
+    fn configure_with(self, config: Config) -> ConfiguredSystemSet;
+    fn before<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSystemSet;
+    fn after<M>(self, label: impl AsSystemLabel<M>) -> ConfiguredSystemSet;
+    fn in_set(self, set: impl SystemLabel) -> ConfiguredSystemSet;
+    fn run_if<P>(self, condition: impl IntoRunCondition<P>) -> ConfiguredSystemSet;
+}
+
+enum Order {
+    Before,
+    After
+}
+
+struct Config {
+    sets: HashSet<SystemLabelId>,
+    dependencies: Vec<(Order, SystemLabelId)>,
+    conditions: Vec<BoxedRunCondition>,
+}
+
+struct ConfiguredSystem {
+    system: BoxedSystem,
+    config: Config,
+    instance_name: SystemLabelId,
+}
+
+struct ConfiguredSystemSet {
+    set: SystemLabelId,
+    config: Config,
+}
+```
+
+### Schedule building errors
+
+When prototyping, you'll likely encounter schedule build errors.
+Most of these will point out a reason why some graph is unsolvable.
+
+There are two kinds of graphs the internal builder checks.
+
+The first kind consists of the `.in_set` relationships.
+These form a graph describing the hierachy of all your system sets.
+The second kind consists of the `.before` and `.after` relationships.
+Those form one or more dependency graphs (one for each set/schedule).
+Dependencies involving system sets will also be flattened into dependencies on the systems in those sets.
+
+So what are the errors?
+
+1. A dependency graph contains a loop or cycle.
+2. The hierarchical graph contains a loop or cycle.
+3. The hierarchical graph has a transitive edge. Example:

```

Comment:
Agreed, I think these should be a `warn!`

Reply:
I misread that though. It refers to the hierarchichal graph, not the intra-set dependency graph. So it may create problems with flattening I guess. So, ideally this should perform transitive reduction and log a warning, but for simplicity reasons it just panics.

Maybe we could add this feature to the future possibilities.

Reply:
Transitive errors in the hierarchical graph are indeed errors.